sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4

```diff
@@ -219,6 +219,12 @@ statement
     | RESET                                                            #resetConfiguration
     | DELETE FROM multipartIdentifier tableAlias whereClause?          #deleteFromTable
     | UPDATE multipartIdentifier tableAlias setClause whereClause?     #updateTable
+    | MERGE INTO target=multipartIdentifier targetAlias=tableAlias
+        USING (source=multipartIdentifier |
+          '(' sourceQuery=query')') sourceAlias=tableAlias
```

Comment:
`query` can have cte like `WITH ... SELECT ...`, should we support it as MERGE source?

If you look at `dmlStatementNoWith`, INSERT INTO doesn't support cte in the input query.

Reply:
nvm, I checked with pgsql and we can have cte in multiple places. Below is a valid SQL:
`with s(i) as (select 1) insert into t with d(i) as (select 2) select i from d`.

We should fix it in Spark later.

Reply:
Yea, I also tried postgres and checked some docs but I didn't find any strong argue why we should/should not support cte in MERGE (MERGE was committed to pg code but later reverted due to some unknown reason). I think there's no bad to support this.

I‘ll to add a case for this syntax.